prov/gni/src/gnix_wait.c

```diff
@@ -30,9 +30,159 @@
  * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
  * SOFTWARE.
  */
-
+#include <stdlib.h>
+#include <signal.h>
 #include "gnix.h"
 #include "gnix_wait.h"
+#include "gnix_nic.h"
+
+/*
+ * Gnix wait progress thread declarations for making sure nic progress
+ * occurs when inside a gnix_wait call
+ */
+
+static pthread_t	gnix_wait_thread;
+static pthread_mutex_t	gnix_wait_mutex = PTHREAD_MUTEX_INITIALIZER;
+static pthread_cond_t	gnix_wait_cond;
+static volatile int	gnix_wait_thread_ready;
+static volatile int	gnix_wait_thread_enabled;
+static volatile int	gnix_wait_thread_busy;
+static volatile int	gnix_wait_refcnt;
+
+/*
+ * It is necessary to have a separate thread making progress in order for the
+ * wait functions to succeed. This version of that thread is designed
+ * to always make progress so we don't hard stall while sitting on fi_wait.
+ */
+static void *__gnix_wait_nic_prog_thread_fn(void *the_arg)
+{
+	int ret = FI_SUCCESS, prev_state;
+	struct gnix_nic *nic1, *nic2;
+	sigset_t  sigmask;
+	gni_return_t status;
+	gni_cq_entry_t cqe;
+
+	GNIX_TRACE(FI_LOG_EP_CTRL, "\n");
+
+	/*
+	 * temporarily disable cancelability while we set up
+	 * some stuff
+	 */
+
+	pthread_setcancelstate(PTHREAD_CANCEL_DISABLE, &prev_state);
+
+	/*
+	 * help out Cray core-spec, say we're not an app thread
+	 * and can be run on core-spec cpus.
+	 */
+	gnix_wait_thread_ready = 1;
+	ret = _gnix_task_is_not_app();
+	if (ret)
+		GNIX_WARN(FI_LOG_EP_CTRL,
+			"_gnix_task_is_not_app call returned %d\n",
+			ret);
+
+	/*
+	 * block all signals, don't want this thread to catch
+	 * signals that may be for app threads
+	 */
+
+	memset(&sigmask, 0, sizeof(sigset_t));
+	ret = sigfillset(&sigmask);
+	if (ret) {
+		GNIX_WARN(FI_LOG_EP_CTRL,
+		"sigfillset call returned %d\n", ret);
+	} else {
+
+		ret = pthread_sigmask(SIG_SETMASK,
+					&sigmask, NULL);
+		if (ret)
+			GNIX_WARN(FI_LOG_EP_CTRL,
+			"pthread_sigmask call returned %d\n", ret);
+	}
+
+	/*
+	 * okay now we're ready to be cancelable.
+	 */
+
+	pthread_setcancelstate(PTHREAD_CANCEL_ENABLE, &prev_state);
+
+	pthread_setcanceltype(PTHREAD_CANCEL_ASYNCHRONOUS, NULL);
+
+try_again:
+	pthread_mutex_lock(&gnix_wait_mutex);
+	if (!gnix_wait_thread_enabled)
+		pthread_cond_wait(&gnix_wait_cond, &gnix_wait_mutex);
+	pthread_mutex_unlock(&gnix_wait_mutex);
+	status = GNI_RC_SUCCESS;
```

Comment:
Good idea with regard to this. We still need the mutex for the pthread_cond_wait but we could just take the mutex after we succeed at entering the if block.